### PR TITLE
Targets to generate macos gen snapshot.

### PR DESCRIPTION
**Pull request by godofredoc**
This is to move the logic to generate the macos snapshots to gn + ninja.

Bug: https://github.com/flutter/flutter/issues/81855

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [X] I listed at least one issue that this PR fixes in the description above.
- [X] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] I signed the [CLA].
- [X] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: common/config.gni
Comment (zanderso) [inline common/config.gni:78]: Mixing the host os with the target cpu doesn't make sense. Only host_os/host_cpu or target_os/target_cpu makes sense.

File: sky/tools/gen_snapshots.py
Comment (zanderso) [inline sky/tools/gen_snapshots.py:1]: How is this different from https://github.com/flutter/engine/blob/main/sky/tools/create_macos_gen_snapshots.py ?

File: common/config.gni
Comment (godofredoc) [reply common/config.gni]: Created an additional variable using target_os and target_cpu which is required for the output name of the dart-sdk and left this one intact.

File: sky/tools/gen_snapshots.py
Comment (godofredoc) [reply sky/tools/gen_snapshots.py:1]: The other script relies on x64, arm64 and armv7 to be build in the same machine which is not always true if running from gn + ninja context. This new script processes a single architecture at a time.

**Comment by godofredoc** [comment]
The PR is ready for a second review.

**Review by zanderso — DISMISSED** [review]
lgtm w/ question

File: shell/platform/embedder/BUILD.gn
Comment (godofredoc) [reply shell/platform/embedder/BUILD.gn:440]: Actually that's a good question, the zip structure is solving the symlinks and copying the files. Would that work?

https://cas-viewer.appspot.com/projects/chromium-swarm/instances/default_instance/blobs/197ebc2f02363983a3e517e3fc1fa1cc72c6cc5fa9db3d983224988ae8d147fb/392/tree

**Comment by chinmaygarde** [comment]
Can we land this?

File: shell/platform/embedder/BUILD.gn
Comment (zanderso) [reply shell/platform/embedder/BUILD.gn:440]: No. The symlink structure needs to be maintained.

Comment (godofredoc) [reply shell/platform/embedder/BUILD.gn:440]: Symlinks are not officially supported in zip files and not implemented in the python zipfile module. I'll need to look for alternatives.

**Comment by chinmaygarde** [comment]
@godofredoc Any updates on available alternatives? In any case, we can close this right?

**Comment by godofredoc** [comment]
> updates on available alternatives? In any case, we can close this right?

We only need to update the zip python script to outsource the creation of the zip file to the zip binary. I'll update that in a separate PR and we should be able to land this one as soon as the dependent PR lands.

File: shell/platform/embedder/BUILD.gn
Comment (godofredoc) [reply shell/platform/embedder/BUILD.gn:440]: This is now fixed. The zip script will be including symlinks correctly.

**Comment by godofredoc** [comment]
> Can we land this?

This is now ready to land if there are no any other concerns. Symlinks have been fixed.

**Review by zanderso — APPROVED** [review]

**Pull request merged**
